File: internal/engine/batch.go
package engine

import (
	"context"
	"os"
	"runtime"
	"sync"
)

// Progress is an immutable snapshot of batch hashing progress. Snapshots are
// delivered to BatchOptions.Progress from worker goroutines; the snapshot
// never aliases internal state and may be retained by the receiver.
//
// Progress follows a small state machine:
//
//   - Starting a file sets CurrentPath and CurrentSize and resets CurrentBytes
//     to zero.
//   - While the file is hashed, CurrentBytes advances as bytes are read.
//     CompletedBytes deliberately excludes the in-flight file, so
//     CompletedBytes+CurrentBytes is the number of bytes hashed so far.
//   - Completing a file, successfully or not, increments CompletedFiles, adds
//     the bytes the attempt read to CompletedBytes and resets CurrentBytes to
//     zero. CurrentPath and CurrentSize keep describing the file that just
//     finished until another file starts.
//
// Files skipped because the batch was cancelled before they started emit no
// events, so CompletedFiles ends below TotalFiles when work is cancelled.
// CurrentSize and TotalBytes are the sizes observed when the batch started;
// files whose size could not be determined contribute zero.
type Progress struct {
	CompletedFiles int
	TotalFiles     int
	CompletedBytes int64
	TotalBytes     int64
	CurrentPath    string
	CurrentBytes   int64
	CurrentSize    int64
}

// BatchOptions configures HashFiles. Workers defaults to min(4, GOMAXPROCS) and
// Progress, when set, may be called concurrently.
type BatchOptions struct {
	HashOptions Options
	Workers     int
	Progress    func(Progress)
	FailFast    bool
}

// FileRequest names one file and the algorithms to compute for it.
type FileRequest struct {
	Path       string
	Algorithms []string
}

// FileResult pairs a FileRequest with its Result or the error that prevented it.
type FileResult struct {
	Request FileRequest
	Result  Result
	Err     error
}

// HashFiles hashes requests concurrently and returns one result per request, in
// request order. FailFast cancels the remaining work after the first failure.
func HashFiles(ctx context.Context, requests []FileRequest, options BatchOptions) []FileResult {
	results := make([]FileResult, len(requests))
	if len(requests) == 0 {
		return results
	}
	workers := options.Workers
	if workers <= 0 {
		workers = min(4, runtime.GOMAXPROCS(0))
	}
	if workers > len(requests) {
		workers = len(requests)
	}

	var totalBytes int64
	fileSizes := make([]int64, len(requests))
	for index, request := range requests {
		if info, err := os.Stat(request.Path); err == nil {
			fileSizes[index] = info.Size()
			totalBytes += info.Size()
		}
	}
	ctx, cancel := context.WithCancel(ctx)
	defer cancel()
	jobs := make(chan int)
	var group sync.WaitGroup
	group.Add(workers)
	tracker := &progressTracker{progress: Progress{TotalFiles: len(requests), TotalBytes: totalBytes}}
	for worker := 0; worker < workers; worker++ {
		go func() {
			defer group.Done()
			for index := range jobs {
				request := requests[index]
				if err := ctx.Err(); err != nil {
					results[index] = FileResult{Request: request, Err: err}
					continue
				}
				emitProgress(options.Progress, tracker.update(func(progress *Progress) {
					progress.CurrentPath = request.Path
					progress.CurrentBytes = 0
					progress.CurrentSize = fileSizes[index]
				}))

				hashOptions := options.HashOptions
				hashOptions.Progress = func(bytesRead int64) {
					emitProgress(options.Progress, tracker.update(func(progress *Progress) {
						progress.CurrentBytes = bytesRead
					}))
				}
				result, err := HashFileWithOptions(ctx, request.Path, request.Algorithms, hashOptions)
				results[index] = FileResult{Request: request, Result: result, Err: err}
				emitProgress(options.Progress, tracker.update(func(progress *Progress) {
					progress.CompletedFiles++
					progress.CompletedBytes += result.BytesRead
					progress.CurrentBytes = 0
				}))
				if err != nil && options.FailFast {
					cancel()
				}
			}
		}()
	}
	for index := range requests {
		if err := ctx.Err(); err != nil {
			results[index] = FileResult{Request: requests[index], Err: err}
			continue
		}
		select {
		case jobs <- index:
		case <-ctx.Done():
			results[index] = FileResult{Request: requests[index], Err: ctx.Err()}
		}
	}
	close(jobs)
	group.Wait()
	return results
}

// progressTracker serializes progress updates so that the shared counters stay
// consistent. The user callback is always invoked after the lock is released:
// a slow or re-entrant callback therefore cannot stall the workers or deadlock
// the batch. Callbacks may run concurrently with one another, so receivers
// must be safe for concurrent use.
type progressTracker struct {
	mu       sync.Mutex
	progress Progress
}

func (tracker *progressTracker) update(mutate func(*Progress)) Progress {
	tracker.mu.Lock()
	mutate(&tracker.progress)
	snapshot := tracker.progress
	tracker.mu.Unlock()
	return snapshot
}

func emitProgress(callback func(Progress), progress Progress) {
	if callback != nil {
		callback(progress)
	}
}

File: src/core/batch.cpp
#include "core/batch.h"

#include <QFileInfo>
#include <QThread>

#include <atomic>
#include <thread>
#include <vector>

namespace hash_core {
namespace {

// progressTracker serializes progress updates so that the shared counters
// stay consistent. The user callback is always invoked after the lock is
// released: a slow or re-entrant callback therefore cannot stall the workers
// or deadlock the batch. Callbacks may run concurrently with one another, so
// receivers must be safe for concurrent use.
class ProgressTracker {
  public:
    Progress update(const std::function<void(Progress &)> &mutate) {
        std::lock_guard<std::mutex> lock(m_mutex);
        mutate(m_progress);
        return m_progress;
    }

  private:
    std::mutex m_mutex;
    Progress m_progress;
};

void emitProgress(const std::function<void(const Progress &)> &callback, const Progress &progress) {
    if (callback) {
        callback(progress);
    }
}

} // namespace

QVector<FileResult> hashFiles(const QVector<FileRequest> &requests, const BatchOptions &options) {
    QVector<FileResult> results(requests.size());
    if (requests.isEmpty()) {
        return results;
    }

    int workers = options.workers;
    if (workers <= 0) {
        const int ideal = QThread::idealThreadCount();
        workers = qMin(4, ideal > 0 ? ideal : 1);
    }
    workers = int(qMin<qint64>(workers, requests.size()));

    qint64 totalBytes = 0;
    QVector<qint64> fileSizes(requests.size(), 0);
    for (int index = 0; index < requests.size(); ++index) {
        const QFileInfo info(requests.at(index).path);
        if (info.exists()) {
            fileSizes[index] = info.size();
            totalBytes += fileSizes.at(index);
        }
    }

    std::atomic<bool> localCancel{false};
    auto canceled = [&localCancel, &options]() {
        return localCancel.load(std::memory_order_relaxed) ||
               (options.canceled && options.canceled());
    };

    ProgressTracker tracker;
    const int totalFiles = requests.size();
    tracker.update([totalFiles, totalBytes](Progress &progress) {
        progress.totalFiles = totalFiles;
        progress.totalBytes = totalBytes;
    });

    std::atomic<int> nextIndex{0};
    auto workerLoop = [&]() {
        for (;;) {
            const int index = nextIndex.fetch_add(1, std::memory_order_relaxed);
            if (index >= requests.size()) {
                break;
            }
            const FileRequest &request = requests.at(index);
            if (canceled()) {
                FileResult &result = results[index];
                result.request = request;
                result.error = QStringLiteral("context canceled");
                result.canceled = true;
                continue;
            }
            emitProgress(options.progress,
                         tracker.update([&request, &fileSizes, index](Progress &progress) {
                             progress.currentPath = request.path;
                             progress.currentBytes = 0;
                             progress.currentSize = fileSizes.at(index);
                         }));

            HashOptions hashOptions = options.hash;
            hashOptions.canceled = canceled;
            hashOptions.progress = [&tracker, &callback = options.progress](qint64 bytesRead) {
                emitProgress(callback, tracker.update([bytesRead](Progress &progress) {
                    progress.currentBytes = bytesRead;
                }));
            };
            const HashOutcome outcome =
                hashFileWithOptions(request.path, request.algorithms, hashOptions);

            FileResult &result = results[index];
            result.request = request;
            result.result = outcome.result;
            result.error = outcome.error;
            result.canceled = outcome.canceled;
            emitProgress(options.progress, tracker.update([&outcome](Progress &progress) {
                progress.completedFiles += 1;
                progress.completedBytes += outcome.result.bytesRead;
                progress.currentBytes = 0;
            }));
            if (!outcome.error.isEmpty() && options.failFast) {
                localCancel.store(true, std::memory_order_relaxed);
            }
        }
    };

    std::vector<std::thread> threads;
    threads.reserve(size_t(workers));
    for (int worker = 0; worker < workers; ++worker) {
        threads.emplace_back(workerLoop);
    }
    for (std::thread &thread : threads) {
        thread.join();
    }
    return results;
}

} // namespace hash_core
